### ISSUE TITLE
Mercurial status is much slower than git status

### ISSUE THREAD

**Issue by azmeuk** [nojhan/liquidprompt#680]
I apologize in advance for the lack of precision and reproducibility of this bug.

I use mercurial AND git on a daily basis for my work. I just noticed that when I work on a mercurial repository with liquidprompt, prompt is displayed a lot more slower than in a git repository, and I have a quite powerful computer, and quite little repositories (<1000 commits).

I am not sure if it is due to liquidprompt or mercurial.

Shell:
ZSH 5.8

Operating system:
Linux ronflex 5.12.15-arch1-1 #1 SMP PREEMPT Wed, 07 Jul 2021 23:35:29 +0000 x86_64 GNU/Linux
LSB Version:	1.4
Distributor ID:	Arch
Description:	Arch Linux
Release:	rolling
Codename:	n/a

Liquidprompt version: 
2.0.2

### Expected Behavior
The prompt displays in a blink of an eye, even of mercurial repositories.

### Current Behavior
The prompt is quite slow to display on mercurial repositories.

**Comment by azmeuk** [comment]
Here is a totally unscientific time measurement of the hg and git commands that liquidprompt does.
We can see that mercurial commands takes a lot more time than git ones :(

```
$ time hg branch
default
hg branch  0,13s user 0,01s system 99% cpu 0,138 total

$ hg bookmark --list --quiet
hg bookmark --list --quiet  0,13s user 0,01s system 99% cpu 0,143 total

$ time hg identify --template='{tags}'
tiphg identify --template='{tags}'  0,15s user 0,04s system 99% cpu 0,184 total

$ time hg identify --id               
14aa606de5ad
hg identify --id  0,16s user 0,02s system 99% cpu 0,186 total

$ time hg shelve --list                 
hg shelve --list  0,15s user 0,02s system 99% cpu 0,169 total

$ time hg status --unknown --template '{status}'               
hg status --unknown --template '{status}'  0,15s user 0,03s system 99% cpu 0,180 total

$ time hg status --modified --template '{status}'
hg status --modified --template '{status}'  0,14s user 0,03s system 99% cpu 0,173 total

$ time hg diff --stat 2>/dev/null | sed -n '$ s/^.*, \([0-9]*\) .*, \([0-9]*\).*$/\1 \2/p'                                         
hg diff --stat 2> /dev/null  0,15s user 0,02s system 84% cpu 0,203 total
sed -n '$ s/^.*, \([0-9]*\) .*, \([0-9]*\).*$/\1 \2/p'  0,01s user 0,00s system 2% cpu 0,203 total
```

```
$ time git rev-parse --is-inside-work-tree                                                                                  
true
git rev-parse --is-inside-work-tree  0,00s user 0,01s system 89% cpu 0,006 total

$ time git symbolic-ref -q HEAD
refs/heads/master
git symbolic-ref -q HEAD  0,00s user 0,01s system 88% cpu 0,006 total

$ time git describe --tags --exact-match                       
fatal: aucune étiquette ne correspond parfaitement à '87599a7408cdfde56c7a80d3fce74f37f32017d0'
git describe --tags --exact-match  0,01s user 0,00s system 74% cpu 0,019 total

$ time git rev-parse -q HEAD                                
87599a7408cdfde56c7a80d3fce74f37f32017d0
git rev-parse -q HEAD  0,00s user 0,01s system 90% cpu 0,006 total

$ time git rev-list --walk-reflogs --count refs/stash                    
fatal: argument 'refs/stash' ambigu : révision inconnue ou chemin inexistant.
Utilisez '--' pour séparer les chemins des révisions, comme ceci :
'git <commande> [<révision>...] -- [<chemin>...]'
git rev-list --walk-reflogs --count refs/stash  0,00s user 0,00s system 90% cpu 0,006 total

$ time git rev-list --count --left-right '@{upstream}...HEAD'                                             
0	0
git rev-list --count --left-right '@{upstream}...HEAD'  0,00s user 0,00s system 95% cpu 0,005 total

$ time LC_ALL=C \git status --porcelain                                                     
 M BTrees/_base.py
LC_ALL=C \git status --porcelain  0,00s user 0,00s system 91% cpu 0,007 total

$ time LC_ALL=C \git diff --shortstat HEAD                               
 1 file changed, 262 insertions(+), 209 deletions(-)
LC_ALL=C \git diff --shortstat HEAD  0,01s user 0,00s system 42% cpu 0,042 total

$ time LC_ALL=C \git diff --shortstat --cached                                  
LC_ALL=C \git diff --shortstat --cached  0,00s user 0,01s system 106% cpu 0,010 total
```

**Comment by Rycieos** [comment]
[Version 2.0.3](https://github.com/nojhan/liquidprompt/releases/tag/v2.0.3) fixed a bug in our handling of `hg`, I doubt it would fix your issue, but you should try it anyway.

I am quite sure that the fault lies with Mercurial. Here is what I did to test this:
```bash
[/tmp] $ hg clone https://www.mercurial-scm.org/repo/hello
destination directory: hello
...
updating to branch default
2 files updated, 0 files merged, 0 files removed, 0 files unresolved
[/tmp] $ cd hello/
[/tmp/hello] default ☿ time _lp_vcs_details_color

real   	0m1.000s
user   	0m0.848s
sys    	0m0.138s
[/tmp/hello] default ☿ cd ~/dev/liquidprompt/
[~/dev/liquidprompt] master+ ± time _lp_vcs_details_color

real   	0m0.027s
user   	0m0.009s
sys    	0m0.020s
```

So Git takes 2.7% of the time Mercurial does.

Running just bare commands, this tracks as well:
```bash
[~/dev/liquidprompt] master+ 130 ± time git symbolic-ref -q HEAD
refs/heads/master

real   	0m0.003s
user   	0m0.000s
sys    	0m0.002s
[/tmp/hello] default ☿ time hg branch
default

real   	0m0.189s
user   	0m0.164s
sys    	0m0.024s
```
It actually is all in the startup time:
```bash
$ time git --version
git version 2.27.0

real   	0m0.002s
user   	0m0.000s
sys    	0m0.002s
$ time hg --version
Mercurial Distributed SCM (version 5.8.1)
...

real   	0m0.177s
user   	0m0.157s
sys    	0m0.019s
```
I think as long as Mercurial is using Python (and Python 2 at that, am I the only one that thinks that Mercurial is already in the grave?) it will always be painfully slow.

That's not to say there are no ways Liquidprompt could improve on that. I did some work for v2.0 to speed up repository detection by using Shell to look for VCS databases. Something similar could be done to rewrite `_lp_hg_branch()` to instead read directly from `.hg/branch`. Something like:
```diff
diff --git a/liquidprompt b/liquidprompt
index f0a1848..fc2fedf 100755
--- a/liquidprompt
+++ b/liquidprompt
@@ -1946,7 +1946,8 @@ _lp_hg_active() {
 # Get the branch name of the Mercurial repo in the current directory.
 _lp_hg_branch() {
     local branch ret
-    if branch="$(\hg branch 2>/dev/null)"; then
+    IFS='' read branch < "${lp_vcs_root}/.hg/branch"
+    if [[ -n $branch ]]; then
         __lp_escape "$branch"
         lp_vcs_branch="$ret"
     else
```
That saves all the time of invoking `hg branch` as there is no forking call. But it has issues, like

* depending on `$lp_vcs_root` being set (and correct).
* depending on the HG database being in the right format.
* who knows what else, since we are doing it ourselves.

And this is only easy for the branch name. All the other bits of info are much more complex and probably not a good idea to do in Shell. And would also bloat Liquidprompt.

In summary, I'm closing this because it is a problem with the tool. But if sensible, well thought out, and edge-case covering patches are submitted, of course I would merge them.

**Comment by azmeuk** [comment]
Ok. Thank you for the detailed answer.

**Comment by azmeuk** [comment]
@Rycieos Would you accept some kind of parallelization for commands, such as described [here](https://stackoverflow.com/a/20018118/2180332)?

**Comment by Rycieos** [comment]
> 
> python alone does not seem to be that slow:

No, it is not. But if you run `strace hg --version`, you will see how many extra system calls and file opens it needs to do just to print the version string.

> 
> 
> Would you accept a patch that brings some kind of parallelization for commands, such as described [here](https://stackoverflow.com/a/20018118/2180332)?

Parallelization has never been tried in Liquidprompt, probably because Shell (like Python) is so bad at it. Ideally there would be functionality like Python's `await` keyword, but I don't think Shell will ever be capable of that.

The best way I have seen to do parallelization in Shell is this:
```bash
for thing in "${things[@]}"; do
    do_thing "$thing" &
done
wait
```

But this doesn't work when the shell is in interactive mode (which is is for Liquidprompt), as the jobs share the same space as user defined jobs. So `wait` would block on any jobs the user started, and all `&` evocations will print the job details on start and stop.

There is a better way, using `coproc`. [See this post](https://unix.stackexchange.com/questions/86270/how-do-you-use-the-command-coproc-in-various-shells/86372#86372) for a amazing breakdown of the feature by the always amazing Stéphane Chazelas. But `coproc` has serious issues:

* It was added in Bash 4.0, but we currently support Bash 3.2+
* It is in both Zsh and Bash, but their implementations are different, complicating usage and reducing usefulness
* It is prone to deadlocking

I would love to parallelize Liquidprompt, but I think I would need to see a proof-of-concept first before I would attempt anything myself. `await` would be the right way to do it, but I just don't think it can exist for Shell. And `coproc` would need serious re-writes to everything to get any speedup from it.

**Comment by azmeuk** [comment]
> See this post for a amazing breakdown of the feature [...]

Thank you for the link!

> It was added in Bash 4.0, but we currently support Bash 3.2+

We can imagine parallelization is supported, and the regular behavior else wise.

I just had a chat with a mercurial developer. They plan to release the version 6, with parts rewritten in Rust, for the end of the year. It should bring better performances. I think I'll just wait and see.

**Comment by Rycieos** [comment]
> 
> 
> They plan to release the version 6, with parts rewritten in Rust, for the end of the year. It should bring better performances.

Very interesting news. I found [this page](https://www.mercurial-scm.org/wiki/OxidationPlan) with their current plans, and they specifically mention shell prompts. I'm glad to see they are taking steps to fix their various issues.

**Comment by azmeuk** [comment]
The recent mercurial 6.1 brings [rhg](https://www.mercurial-scm.org/repo/hg/file/tip/rust/README.rst) but there is not enough features to be usable yet. However, I just stumbled upon [chg](https://www.mercurial-scm.org/wiki/CHg) which really improves perfs!
